sklearn/utils/tests/test_testing.py

```diff
@@ -211,26 +211,20 @@ def context_manager_no_user_multiple_warning():
     assert_warns(DeprecationWarning, context_manager_no_user_multiple_warning)
 
 
-# This class is inspired from numpy 1.7 with an alteration to check
-# the reset warning filters after calls to assert_warns.
-# This assert_warns behavior is specific to scikit-learn because
-# `clean_warning_registry()` is called internally by assert_warns
-# and clears all previous filters.
+# assert_warns will no longer clear all previous filters.
 class TestWarns(unittest.TestCase):
     def test_warn(self):
         def f():
             warnings.warn("yo")
             return 3
 
-        # Test that assert_warns is not impacted by externally set
-        # filters and is reset internally.
-        # This is because `clean_warning_registry()` is called internally by
-        # assert_warns and clears all previous filters.
+        # test that assert_warns doesn't affect external filters
+        warnings.resetwarnings()
```

Comment:
Please do not mess with the global `warnings.filters`. Use a `warnings.catch_warnings()` context manager if you want to use `warwnings.resetwarnings`, `warning.simplefilter` or `warnings.filterwarnings`.

Basically do:
```py
with warnings.catch_warnings():
    warnings.simplefilter('ignore', UserWarning)
    filters_orig = warnings.filters[:]
    assert_equal(assert_warns(UserWarning, f), 3)
    assert_equal(warnings.filters, filters_orig)
```

I added something that checks that the warnings.filters are not affected by assert_warns (not tested)